### PR TITLE
[WASI-Socket] send_to support

### PR DESCRIPTION
**Pull request by LFsWang**
Test reference : https://chromium.googlesource.com/chromium/src/+/refs/heads/main/net/socket/udp_socket_unittest.cc#288

### CODE REVIEW

File: lib/host/wasi/inode-linux.cpp [hunk 1]
```diff
@@ -1037,6 +1037,38 @@ WasiExpect<void> INode::sockSend(Span<Span<const uint8_t>> SiData,
...
+  msghdr SysMsgHdr;
+  SysMsgHdr.msg_name = Address;
+  SysMsgHdr.msg_namelen = AddressLength;
+  SysMsgHdr.msg_iov = SysIOVs;
+  SysMsgHdr.msg_iovlen = SysIOVsSize;
+  SysMsgHdr.msg_control = nullptr;
+  SysMsgHdr.msg_controllen = 0;
+
+  /// Store recv bytes length and flags.
+  if (auto Res = ::sendmsg(Fd, &SysMsgHdr, SysSiFlags); unlikely(Res < 0)) {
```
Comment (hydai) [inline lib/host/wasi/inode-linux.cpp:1063]: Is this another function?

**Review by hydai — CHANGES_REQUESTED** [review]

File: lib/host/wasi/inode-linux.cpp (hunk 1 above)
Comment (LFsWang) [reply lib/host/wasi/inode-linux.cpp:1063]: By linux document, ```send(...) = sendto(..., NULL, 0) = sendmsg( Msg{...} ) ```, therefore I use sendmsg as same as the implementation for send ```send()``` 

https://linux.die.net/man/2/sendto

Comment (LFsWang) [reply lib/host/wasi/inode-linux.cpp:1063]: In other words, `sendto` = `send` + `destination address buffer`

**Review by ibmibmibm — DISMISSED** [review]

**Comment by hydai** [comment]
Hi @LFsWang 
Your DCO sign is not available. Please rebase and sign your commit with DCO. `git commit --sign-off`

**Review by hydai — APPROVED** [review]

File: lib/host/wasi/inode-linux.cpp [hunk 2]
```diff
@@ -1022,14 +1037,14 @@ WasiExpect<void> INode::sockSend(Span<Span<const uint8_t>> SiData,
...
-  SysMsgHdr.msg_namelen = 0;
+  SysMsgHdr.msg_name = Address;
+  SysMsgHdr.msg_namelen = AddressLength;
   SysMsgHdr.msg_iov = SysIOVs;
   SysMsgHdr.msg_iovlen = SysIOVsSize;
   SysMsgHdr.msg_control = nullptr;
   SysMsgHdr.msg_controllen = 0;
 
-  // Store recv bytes length and flags.
+  /// Store recv bytes length and flags.
```
Comment (hydai) [inline lib/host/wasi/inode-linux.cpp:1047]: Comments is `//` only.

File: lib/host/wasi/inode-macos.cpp [hunk 3]
```diff
@@ -902,14 +917,14 @@ WasiExpect<void> INode::sockSend(Span<Span<const uint8_t>> SiData,
...
-  SysMsgHdr.msg_namelen = 0;
+  SysMsgHdr.msg_name = Address;
+  SysMsgHdr.msg_namelen = AddressLength;
   SysMsgHdr.msg_iov = SysIOVs;
   SysMsgHdr.msg_iovlen = SysIOVsSize;
   SysMsgHdr.msg_control = nullptr;
   SysMsgHdr.msg_controllen = 0;
 
-  // Store recv bytes length and flags.
+  /// Store recv bytes length and flags.
```
Comment (hydai) [inline lib/host/wasi/inode-macos.cpp:927]: Ditto.

File: lib/host/wasi/wasifunc.cpp [hunk 4]
```diff
@@ -1834,6 +1921,90 @@ Expect<uint32_t> WasiSockSend::body(Runtime::Instance::MemoryInstance *MemInst,
   return __WASI_ERRNO_SUCCESS;
 }
 
+Expect<uint32_t>
+WasiSockSendTo::body(Runtime::Instance::MemoryInstance *MemInst, int32_t Fd,
+                     uint32_t SiDataPtr, uint32_t SiDataLen,
+                     uint32_t AddressPtr, uint32_t SiFlags,
+                     uint32_t /* Out */ SoDataLenPtr) {
+  /// Check memory instance from module.
```
Comment (hydai) [inline lib/host/wasi/wasifunc.cpp:1929]: Ditto.

File: lib/host/wasi/wasifunc.cpp [hunk 5]
```diff
@@ -1834,6 +1921,90 @@ Expect<uint32_t> WasiSockSend::body(Runtime::Instance::MemoryInstance *MemInst,
...
+  } else {
+    WasiSiFlags = *Res;
+  }
+
+  const __wasi_size_t WasiSiDataLen = SiDataLen;
+  if (unlikely(WasiSiDataLen > WASI::kIOVMax)) {
+    return __WASI_ERRNO_INVAL;
+  }
+
+  /// Check for invalid address.
```
Comment (hydai) [inline lib/host/wasi/wasifunc.cpp:1962]: Ditto.

File: lib/host/wasi/wasifunc.cpp [hunk 6]
```diff
@@ -1834,6 +1921,90 @@ Expect<uint32_t> WasiSockSend::body(Runtime::Instance::MemoryInstance *MemInst,
...
+    return __WASI_ERRNO_FAULT;
+  }
+
+  __wasi_size_t TotalSize = 0;
+  std::array<Span<const uint8_t>, WASI::kIOVMax> WasiSiData;
+
+  for (__wasi_size_t I = 0; I < WasiSiDataLen; ++I) {
+    __wasi_ciovec_t &SiData = SiDataArray[I];
+
+    /// Capping total size.
```
Comment (hydai) [inline lib/host/wasi/wasifunc.cpp:1980]: Ditto.

**Review by hydai — CHANGES_REQUESTED** [review]

**Review by hydai — APPROVED** [review]

**Pull request merged**